include/dd/Node.hpp

```diff
@@ -28,20 +32,181 @@ namespace dd {
+                setFirstEdgeDensityPathTrueNode(p->e[1].p);
+
+                for (auto& edge: p->e) {
+                    setDensityMatrixTrueNode(edge.p);
+                }
+
+            } else {
+                std::swap(p->e[2], p->e[1]);
+                for (auto& edge: p->e) {
+                    // Conjugate all edges
+                    edge.w.i = dd::CTEntry::flipPointerSign(edge.w.i);
+                    setDensityConjugateTrueNode(edge.p);
+                    setDensityMatrixTrueNode(edge.p);
+                }
+            }
+            return flags;
+        }
+
+        static inline void getAlignedNodeRevertModificationsOnSubEdges(dNode* p) {
+            // Before I do anything else, I must align the pointer
+            getAlignedDensityNode(p);
+
+            for (auto& edge: p->e) {
+                // first edge paths are not modified I only have to remove the first edge property
+                getAlignedDensityNode(edge.p);
+            }
+
+            if (isFirstEdgeDensityPath(p->flags) && !isDensityConjugateSet(p->flags)) {
+                ;
+
+            } else if (!isDensityConjugateSet(p->flags)) {
+                // Conjugate the second edge (i.e. negate the complex part of the second edge)
+                p->e[2].w.i = dd::CTEntry::flipPointerSign(p->e[2].w.i);
+
+            } else {
+                for (auto& edge: p->e) {
+                    // Align all nodes and conjugate the weights
+                    edge.w.i = dd::CTEntry::flipPointerSign(edge.w.i);
+                }
+                std::swap(p->e[2], p->e[1]);
+            }
+        }
+
+        static inline void applyDmChangesToNode(dNode*& p) {
+            // Align the node pointers
+            if (isDensityMatrix(p)) {
+                auto tmp = alignDensityNodeNode(p);
+                assert(getTempDensityMatrixFlags(p->flags) == 0);
+                p->flags = p->flags | tmp;
+            }
+        }
+
+        static inline void revertDmChangesToNode(dNode*& p) {
+            // Align the node pointers
+            if (isDensityMatrix(p->flags)) {
+                getAlignedNodeRevertModificationsOnSubEdges(p);
+                p->unsetTempDensityMatrixFlags();
+            }
+        }
+    };
+    using dEdge       = Edge<dNode>;
+    using dCachedEdge = CachedEdge<dNode>;
+    inline dNode dNode::terminalNode{{{{nullptr, Complex::zero}, {nullptr, Complex::zero}, {nullptr, Complex::zero}, {nullptr, Complex::zero}}}, nullptr, 0, -1, 0};
+
 } // namespace dd
+
+namespace std {
+    template<>
+    struct hash<dd::dEdge> {
+        std::size_t operator()(dd::dEdge const& e) const noexcept {
+            auto h1 = dd::murmur64(reinterpret_cast<std::size_t>(e.p));
+            auto h2 = std::hash<dd::Complex>{}(e.w);
+            assert((dd::dNode::isDensityMatrix((long)e.p)) == false);
+            auto h3  = std::hash<std::uint_least8_t>{}(dd::dNode::getTempDensityMatrixFlags(e.p->flags));
```

Comment:
I still don't see how this hash makes sense/improves upon the general hash used for Edges.

Assume that there are two `dEdge`s `e` and `f`, and it holds that `e.w` == f.w`.
Then, either `e.p == f.p` (pointers are equal) or `e.p != f.p` (pointers are not equal).
In the first case, since both pointers are identical, it is guaranteed that `e.p->flags == f.p->flags` and, hence, both `e` and `f` would get the same hash. They would also get identical hashes using the original hash function.
In the second case, the hashes are different anyways.

In addition, this has function is rather problematic should `e.p` ever be `nullptr` which would cause an exception.

Reply:
The flag value can change between calls. When the same node has different flag values, this should be represented in the hash. 
Added an assert `assert(e.p != nullptr);`